src/build.d

```diff
@@ -1178,12 +1182,23 @@ class DependencyRef
         if (msg)
             msg.writeln;
 
-        if (commandFunction !is null)
-            return commandFunction();
-
-        if (command)
+        if(dryRun)
+        {
+            if(commandFunction)
```

Comment:
You could try a set of criteria on what to print.  For example, if `name` is null, you could check if there are any outputs and print `generating %s` instead.

Reply:
We could also just make it an error if `commandFunction` is set without `msg` field being set.

Reply:
Changed to list the dependencies `name` and `targets` and fall back to `Executing commandFunction()`.

I think every dependency **must** have a `msg` to reliable track why a build failed.